src/main/java/com/thealgorithms/datastructures/heaps/LeonardoHeap.java

```diff
@@ -0,0 +1,173 @@
+ */
+public class LeonardoHeap<T extends Comparable<T>> {
+
+    private int levelTracker = 0;
+    private final List<T> heap = new ArrayList<T>();
+
+    public LeonardoHeap() {
+    }
+
+    private void decreaseLevelTracker() {
+        int lastTreeLevel = getRightMostTree();
+        levelTracker = SingleBitOperations.clearBit(levelTracker, lastTreeLevel);
+        if (lastTreeLevel != 0 && lastTreeLevel != 1) {
+            levelTracker = SingleBitOperations.setBit(levelTracker, lastTreeLevel - 1);
+            levelTracker = SingleBitOperations.setBit(levelTracker, lastTreeLevel - 2);
+        }
+    }
+
+    private void increaseLevelTracker() {
+        ArrayList<Integer> consecutiveTreeIndices = LeonardoHeapHelper.findConsecutiveTreeIndices(levelTracker);
+        if (consecutiveTreeIndices.get(0) != -1) {
+            // if 0th or 1st index is -1 that implies there are no concequtive trees
+            levelTracker = SingleBitOperations.clearBit(levelTracker, consecutiveTreeIndices.get(0));
+            levelTracker = SingleBitOperations.clearBit(levelTracker, consecutiveTreeIndices.get(1));
+            levelTracker = SingleBitOperations.setBit(levelTracker, consecutiveTreeIndices.get(1) + 1);
+        } else if ((levelTracker & 2) == 0) {
+            levelTracker = SingleBitOperations.setBit(levelTracker, 1);
+        } else {
+            levelTracker = SingleBitOperations.setBit(levelTracker, 0);
+        }
+    }
+
+    private void maxHeapifyTree(int rootNodeIndex, int currentLevel) {
+        // A leonardo tree of level n is just 1 node(the root) plus the leonardo tree of n-1 level(left child) plus leonardo tree of n-2 level(right child)
+        // To maxheapify a leonardo tree we need to compare the current root and roots of it's left and right subtree
+        // We recursively hepify the left and right subtrees using the currentLevel
+
+        // BASE CASE
+        if (currentLevel == 0 || currentLevel == 1) {
+            return; // Trees with one node are in already max-heapified.
+        }
+
+        int currentRootNodeIndex = rootNodeIndex;
+        int rightChildIndex = rootNodeIndex - 1;
+        int leftChildIndex = rootNodeIndex - LeonardoNumber.leonardoNumber(currentLevel - 2) - 1;
+        int childIndexForSwap = -1;
+
+        if (heap.get(rightChildIndex).compareTo(heap.get(leftChildIndex)) >= 0) {
+            childIndexForSwap = rightChildIndex;
+        } else {
+            childIndexForSwap = leftChildIndex;
+        }
+
+        if (heap.get(childIndexForSwap).compareTo(heap.get(currentRootNodeIndex)) > 0) {
+            swap(currentRootNodeIndex, childIndexForSwap);
+            if (childIndexForSwap == rightChildIndex) {
+                maxHeapifyTree(rightChildIndex, currentLevel - 2);
+            } else { // swap happened with the left child
+                maxHeapifyTree(leftChildIndex, currentLevel - 1);
+            }
+        }
+    }
+
+    private void shiftRootAndRestoreHeap() {
```

Comment:
Do you see some change to _chop_ this method into few smaller ones?

Reply:
Might be possible, I'll take a look into using some form of recursion instead of the while loop I am using. Recursion might give a bit more readability.

Reply:
I would suggest to _refactor_ this method into a shorter one.